### PR TITLE
[EBPF-589] Initialize process monitor as a module

### PR DESCRIPTION
**Pull request by gjulianm**
### What does this PR do?

This PR moves the initialization of `ProcessMonitor` to a single module, instead of each module initializing it by itself. It also introduces a new variable for the `module.Factory` called `IgnoreForSuccessCheck`, which ensures that system-probe will still fail initialization if other modules fail but the new process-monitor module is succesful.

### Motivation

The `ProcessMonitor` object is a singleton, but needs to be initialized after all callbacks have been registered so that they receive events for existing processes. Since we have added the GPU monitoring and Dynamic Instrumentation modules which use the process monitor, it can no longer be initialized in the network tracer module. Another problem was that each module would try to initialize the process monitor with different options, causing inconsistencies there.

This is a temporary solution while we plan and develop a better way for system-probe to deal with these dependencies in a generic manner that guarantees proper ordering and initialization only when needed.

https://datadoghq.atlassian.net/browse/EBPF-589

### Describe how to test/QA your changes

Tested locally by:

* starting system probe with USM, GPU, DI enabled and checking that the process monitor is initialized after their callbacks are registered (checking logs)
* Forcing a failure in all modules except the process monitor and checking that system-probe stops.

### Additional Notes

We have not introduced a new config to control the event stream, instead using the existing USM config (`service_monitoring_config.enable_event_stream`) despite that not being used only in USM anymore. A proper configuration option should be defined once we get a better system for managing these dependencies.

### CODE REVIEW

File: cmd/system-probe/modules/processmonitor.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	pkgconfigsetup "github.com/DataDog/datadog-agent/pkg/config/setup"
+	"github.com/DataDog/datadog-agent/pkg/process/monitor"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"github.com/DataDog/datadog-agent/cmd/system-probe/api/module"
+)
+
+func isEventStreamEnabled() bool {
+	eventStreamUSM := pkgconfigsetup.SystemProbe().GetBool("service_monitoring_config.enable_event_stream")
+	eventStreamProcMon := pkgconfigsetup.SystemProbe().GetBool("process_monitor_config.enable_event_stream")
```
Comment (brycekahle) [inline cmd/system-probe/modules/processmonitor.go:25]:
```suggestion
	eventStreamProcMon := pkgconfigsetup.SystemProbe().GetBool("process_monitoring_config.enable_event_stream")
```

**Review by cimi — APPROVED** [review]
:+1: Approving the DI/debugger changes, thank you. The code looks good,  please let us know if you want us to test this build. AFAIK we are now running DI tests in KMT so if those pass we shouldn't need to do manual testing.

Also, we've not tested the process monitor with the event stream on (we were calling `pt.pm.Initialize(false)`) but I assume this should work.

File: cmd/system-probe/api/module/factory_linux.go [hunk 2]
```diff
@@ -17,4 +17,8 @@ type Factory struct {
 	ConfigNamespaces []string
 	Fn               func(cfg *sysconfigtypes.Config, deps FactoryDependencies) (Module, error)
 	NeedsEBPF        func() bool
+
+	// IgnoreForSuccessCheck can be set to true if system-probe should not consider this module when checking
+	// if at least one module was successfully loaded.
+	IgnoreForSuccessCheck bool
```
Comment (guyarb) [inline cmd/system-probe/api/module/factory_linux.go:23]: 1. I think we should rename this variable
2. I don't think you should tie together the new module and this change in a single PR

File: pkg/dynamicinstrumentation/proctracker/proctracker.go [hunk 3]
```diff
@@ -21,17 +21,19 @@ import (
 
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 
-	"github.com/DataDog/datadog-agent/pkg/dynamicinstrumentation/ditypes"
 	"github.com/cilium/ebpf"
 	"github.com/cilium/ebpf/link"
 
+	"github.com/DataDog/datadog-agent/pkg/dynamicinstrumentation/ditypes"
```
Comment (guyarb) [inline pkg/dynamicinstrumentation/proctracker/proctracker.go:27]: should be moved to the last group

File: cmd/system-probe/modules/processmonitor.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	pkgconfigsetup "github.com/DataDog/datadog-agent/pkg/config/setup"
+	"github.com/DataDog/datadog-agent/pkg/process/monitor"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"github.com/DataDog/datadog-agent/cmd/system-probe/api/module"
+)
+
+func isEventStreamEnabled() bool {
+	return pkgconfigsetup.SystemProbe().GetBool("service_monitoring_config.enable_event_stream")
+}
```
Comment (guyarb) [inline cmd/system-probe/modules/processmonitor.go:25]: we have a module that relies on a USM configuration?

File: cmd/system-probe/modules/processmonitor.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	NeedsEBPF: func() bool {
+		return isEventStreamEnabled()
+	},
+	IgnoreForSuccessCheck: true,
+}
+
+var processMonitorModuleConfigNamespaces = []string{"service_monitoring_config"}
+
+func createProcessMonitorModule(_ *sysconfigtypes.Config, _ module.FactoryDependencies) (module.Module, error) {
+	log.Infof("Initializing process monitor...")
```
Comment (guyarb) [inline cmd/system-probe/modules/processmonitor.go:41]: Do we need that log? the module runner should log that for every module

File: cmd/system-probe/api/module/loader.go [hunk 6]
```diff
@@ -122,6 +127,8 @@ func Register(cfg *sysconfigtypes.Config, httpMux *mux.Router, factories []Facto
 	l.cfg = cfg
 	if len(l.modules) == 0 {
 		return errors.New("no module could be loaded")
+	} else if !didAnyModuleSucceed {
+		return errors.New("no required module could be loaded successfully")
```
Comment (guyarb) [inline cmd/system-probe/api/module/loader.go:131]: what's the reasoning behind this change?

**Review by guyarb — CHANGES_REQUESTED** [review]

File: cmd/system-probe/api/module/loader.go (hunk 6 above)
Comment (gjulianm) [reply cmd/system-probe/api/module/loader.go:131]: System-probe should fail to init if no module starts. However, if we add the process monitor module, we will lose that behavior as that module is likely going to be started anyways. That's also why I didn't move that change to another PR as it's very coupled with this module.

**Comment by guyarb** [comment]
Why do you use USM process monitor? Why don't you move directly to eventmonitoring?

File: cmd/system-probe/api/module/factory_linux.go (hunk 2 above)
Comment (gjulianm) [reply cmd/system-probe/api/module/factory_linux.go:23]: This change is very tightly coupled with the rest of the PR, and as it was small I preferred to put it here. Regarding the name, I agree it's not the best but I couldn't thing of anything better.

File: cmd/system-probe/modules/processmonitor.go (hunk 4 above)
Comment (gjulianm) [reply cmd/system-probe/modules/processmonitor.go:25]: For now, yes. The only user-facing setting that controls the usage of event stream is that config, so we have to respect it in this module. Once we do the refactor supporting dependencies of modules we should deprecate the USM config and create a common one.

File: cmd/system-probe/modules/processmonitor.go (hunk 5 above)
Comment (gjulianm) [reply cmd/system-probe/modules/processmonitor.go:41]: You're right, removed.

**Comment by gjulianm** [comment]
> Why do you use USM process monitor? Why don't you move directly to eventmonitoring?

As far as I understand, the `ProcessMonitor` type encapsulates both the netlink and event stream methods of getting process start/exit events, no?

**Comment by guyarb** [comment]
> > Why do you use USM process monitor? Why don't you move directly to eventmonitoring?
> 
> As far as I understand, the `ProcessMonitor` type encapsulates both the netlink and event stream methods of getting process start/exit events, no?

It does, but process monitor is not an improvement, but a legacy component
From the USM perspective, the process monitor (as a netlink wrapper) exists before the event stream monitoring project started.
We didn't adapt event monitoring back then as it didn't fit USM needs (it had issues around consistency and accuracy of events).
Only after all issues have been fixed we started our migration, and as a first phase - add the implementation to process monitor to limit the changes in usm.
Eventually, we will adapt event monitoring fully and remove (or disable) netlink parts

**Review by robertjli — COMMENTED** [review]
It looks like the Processes team is tagged as a codeowner due to this rule
https://github.com/DataDog/datadog-agent/blob/3527f138432ecf119837add93a619e0bd85f04af/.github/CODEOWNERS#L231

Would you have the bandwidth to update it to only `process.go` instead of `process*`? No problem if you don't get to it, I can follow up with a separate PR.

**Comment by gjulianm** [comment]
> It looks like the Processes team is tagged as a codeowner due to this rule
> 
> https://github.com/DataDog/datadog-agent/blob/3527f138432ecf119837add93a619e0bd85f04af/.github/CODEOWNERS#L231
> 
> Would you have the bandwidth to update it to only `process.go` instead of `process*`? No problem if you don't get to it, I can follow up with a separate PR.

Sorry about this, we're probably going to close this PR due to other concerns and follow up with a different approach, I forgot to notify your team about that.

File: cmd/system-probe/api/module/factory_others.go [hunk 7]
```diff
@@ -16,4 +16,8 @@ type Factory struct {
 	Name             sysconfigtypes.ModuleName
 	ConfigNamespaces []string
 	Fn               func(cfg *sysconfigtypes.Config, deps FactoryDependencies) (Module, error)
+
+	// IgnoreForSuccessCheck can be set to true if system-probe should not consider this module when checking
+	// if at least one module was successfully loaded.
+	IgnoreForSuccessCheck bool
```
Comment (yuri-lipnesh) [inline cmd/system-probe/api/module/factory_others.go:22]:
```suggestion
	// SkipModuleCheck - set to 'true' if the system-probe should exclude this module when verifying 
	// whether at least one module was successfully loaded.
	SkipModuleCheck bool
```

**Comment by gjulianm** [comment]
Closing, as we have decided to remove the coupling between uprobe-attacher and process-monitor and use eventstream in DI/GPU, which will solve this problem.

**Pull request closed without merging**